torchtext/vocab.py

```diff
@@ -139,33 +67,184 @@ def __init__(self, counter, max_size=None, min_freq=1, wv_dir=os.getcwd(),
+              expand_vocab (bool): expand vocabulary to include all words for which
+                  the specified pretrained word vectors are available
+        """
+        if not isinstance(vectors, list):
+            vectors = [vectors]
+        vecs = []
+        tot_dim = 0
+        for v in vectors:
+            wv_type = v.split('.')[0]
+            wv_dim = int(v.split('.')[-1][:-1])
+            if wv_type == 'glove':
+                wv_name = '.'.join(v.split('.')[1:-1])
+                vecs.append(GloVe(name=wv_name, dim=wv_dim, unk_init=unk_init))
+                if expand_vocab:
+                    for w in sorted(vecs[-1].stoi.keys()):
+                        self.itos.append(w)
+                        self.stoi[w] = len(self.itos) - 1
+            elif 'charngram' in v:
+                vecs.append(CharNGram(unk_init=unk_init))
+            tot_dim += wv_dim
+
+        self.vectors = torch.Tensor(len(self), tot_dim)
+        start_dim = 0
+        for i, token in enumerate(self.itos):
+            for i, v in enumerate(vectors):
+                end_dim = start_dim + vecs[i].dim
+                self.vectors[i][start_dim:end_dim] = vecs[i][token]
+                start_dim = end_dim
+            assert(start_dim == tot_dim)
+            start_dim = 0
+
+    def set_vectors(self, stoi, vectors, dim, unk_init='zero'):
+        self.vectors = torch.Tensor(len(self), dim)
+        self.vectors.normal_(0, 1) if unk_init == 'random' else self.vectors.zero_()
+        for i, token in enumerate(self.itos):
+            wv_index = stoi.get(token, None)
+            if wv_index is not None:
+                self.vectors[i] = vectors[wv_index]
 
-        Arguments:
-            wv_dir: directory containing word vector file and destination for
-                downloaded word vector files
-            wv_type: type of word vectors; None for no word vectors
-            wv_dim: dimension of word vectors
 
-            unk_init: default to random initialization for unknown word vectors;
-                otherwise set to zero
-        """
+class Vectors(object):
+
+    def __init__(self, unk_init='zero'):
         self.unk_init = unk_init
-        wv_dict, wv_arr, self.wv_size = load_word_vectors(wv_dir, wv_type, wv_dim)
-        self.set_vectors(wv_dict, wv_arr)
 
-    def set_vectors(self, wv_dict, wv_arr):
-        self.vectors = torch.Tensor(len(self), self.wv_size)
-        self.vectors.normal_(0, 1) if self.unk_init == 'random' else self.vectors.zero_()
-        for i, token in enumerate(self.itos):
-            wv_index = wv_dict.get(token, None)
-            if wv_index is not None:
-                self.vectors[i] = wv_arr[wv_index]
+    def __getitem__(self, token):
+        if token in self.stoi:
+            return self.vectors[self.stoi[token]]
+        else:
+            vector = torch.Tensor(1, self.dim).zero_()
+            if self.unk_init == 'random':
+                vector.normal_(0, 1)
```

Comment:
what's the conceptual reason we need to have two places where unk initialization happens?